nav2_costmap_2d/plugins/plugin_container_layer.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2024 Polymath Robotics, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/plugin_container_layer.hpp"
+
+#include "nav2_costmap_2d/costmap_math.hpp"
+#include "nav2_costmap_2d/footprint.hpp"
+#include "nav2_util/node_utils.hpp"
+#include "rclcpp/parameter_events_filter.hpp"
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(nav2_costmap_2d::PluginContainerLayer, nav2_costmap_2d::Layer)
+
+using std::vector;
+
+namespace nav2_costmap_2d
+{
+PluginContainerLayer::PluginContainerLayer()
+{
+}
+
+PluginContainerLayer::~PluginContainerLayer()
+{
+}
+
+void PluginContainerLayer::onInitialize()
+{
+  auto node = node_.lock();
+
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "enabled",
+      rclcpp::ParameterValue(true));
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "plugins",
+      rclcpp::ParameterValue(std::vector<std::string>{}));
+  nav2_util::declare_parameter_if_not_declared(node, name_ + "." + "combination_method",
+      rclcpp::ParameterValue(1));
+
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+  node->get_parameter(name_ + "." + "plugins", plugin_names_);
+
+  int combination_method_param{};
+  node->get_parameter(name_ + "." + "combination_method", combination_method_param);
+  combination_method_ = combination_method_from_int(combination_method_param);
+
+  dyn_params_handler_ = node->add_on_set_parameters_callback(
+    std::bind(
+      &PluginContainerLayer::dynamicParametersCallback,
+      this,
+      std::placeholders::_1));
+
+  plugin_types_.resize(plugin_names_.size());
+
+  for (unsigned int i = 0; i < plugin_names_.size(); ++i) {
```

Comment:
We need to add a try/catch around https://github.com/ros-navigation/navigation2/blob/684a4d037241674550b756c5014711fa6c13e35e/nav2_costmap_2d/src/costmap_2d_ros.cpp#L180-L182 now if we do this, since this can throw

Reply:
done